runtime/v1/shim/client/client.go

```diff
@@ -324,21 +324,33 @@ func (c *Client) signalShim(ctx context.Context, sig syscall.Signal) error {
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
-	case <-c.waitForExit(pid):
+	case <-c.waitForExit(ctx, pid):
 		return nil
 	}
 }
 
-func (c *Client) waitForExit(pid int) <-chan struct{} {
+func (c *Client) waitForExit(ctx context.Context, pid int) <-chan struct{} {
 	c.exitOnce.Do(func() {
+		ctxWithTimeout, cancelFn := context.WithTimeout(ctx, 10*time.Second)
+		defer cancelFn()
+
+		ticker := time.NewTicker(10 * time.Millisecond)
+		defer ticker.Stop()
+
 		for {
-			// use kill(pid, 0) here because the shim could have been reparented
-			// and we are no longer able to waitpid(pid, ...) on the shim
-			if err := unix.Kill(pid, 0); err == unix.ESRCH {
+			select {
+			case <-ticker.C:
```

Comment:
have a funny feeling about waiting the first tick (10mil) before sending the signal from the request

Reply:
Sorry for late reply.

I would like to use `time.Sleep` here, like

```
func (c *Client) waitForExit(ctx context.Context, pid int) <-chan struct{} {
        // use goroutine here
        // otherwise it will be lock until timeout or the shim is gone.
        go c.exitOnce.Do(func() {
                defer close(c.exitCh)

                for {
                        select {
                        case <-ctx.Done():
                                log.G(ctx).WithField("pid", pid).Warn("timed out while waiting for shim to exit")
                                return
                        default:
                        }
                        // use kill(pid, 0) here because the shim could have been reparented
                        // and we are no longer able to waitpid(pid, ...) on the shim
                        if err := unix.Kill(pid, 0); err == unix.ESRCH {
                                return
                        }
                        time.Sleep(10 * time.Millisecond)
                }
        })
        return c.exitCh
}
```

WDYT?

Reply:
@fuweid do you mind pointing to the place where the timeout is being added? 

I also think it's generally harder to reason about a timeout that is being passed in from the CRI plugin here. I'm not sure it's a good idea to rely on that.

Reply:
Sure. It is always 10s https://github.com/containerd/cri/blob/master/pkg/server/events.go#L240.

Yeah. It is hard to make decision about value of timeout. But it is good to let the caller decide it. Maybe the caller uses 60s :). It is tough.

Reply:
is the initial delay necessary?

Reply:
he's using 10ms for the ticker timeout.. before sending the signal

Reply:
> is the initial delay necessary?
@mikebrow no :)

Reply:
suggest something like:
```
func (c *Client) waitForExit(ctx context.Context, pid int) <-chan struct{} {
	go c.exitOnce.Do(func() {		 	

		// use kill(pid, 0) here because the shim could have been reparented
		// and we are no longer able to waitpid(pid, ...) on the shim
		if err := unix.Kill(pid, 0); err == unix.ESRCH {
			close(c.exitCh)
			return
		}

		ctxWithTimeout, cancelFn := context.WithTimeout(ctx, 10*time.Second)
		defer cancelFn()
  
		ticker := time.NewTicker(10 * time.Millisecond)
		defer ticker.Stop()

		for {
			select {
			case <-ticker.C:
				if err := unix.Kill(pid, 0); err == unix.ESRCH {
					close(c.exitCh)
		 			return
		  		}
	 		case <-ctxWithTimeout.Done():
	   			log.G(ctx).WithField("pid", pid).Warn("timed out while waiting for shim to exit")
	   			close(c.exitCh)
	   			return		 				
			}
	  	}
	})
  	return c.exitCh		 
}
```